<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="hyprcoloc">
<title>HyPrColoc • hyprcoloc</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="HyPrColoc">
<meta property="og:description" content="hyprcoloc">
<meta property="og:image" content="https://jrs95.github.io/hyprcoloc/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">hyprcoloc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/hyprcoloc.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jrs95/hyprcoloc/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>HyPrColoc</h1>
                        <h4 data-toc-skip class="author">Christopher N
Foley &amp; James R Staley</h4>
            
            <h4 data-toc-skip class="date">2024-04-05</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jrs95/hyprcoloc/blob/HEAD/vignettes/hyprcoloc.Rmd" class="external-link"><code>vignettes/hyprcoloc.Rmd</code></a></small>
      <div class="d-none name"><code>hyprcoloc.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>HyPrColoc is Bayesian divisive clustering algorithm for identifying
clusters of traits which colocalize at distinct causal variants in a
genomic region. The default algorithm can identify clusters of
putatively colocalized traits within a vast collection of traits,
e.g. 1000s, quickly. For each set of putatively colocalized traits, the
algorithm outputs: the names of the traits, the posterior probability of
colocalization, the location of the shared putatively causal variant and
a ‘fine-mapping’ probability to quantify evidence supporting the
candidate causal variant being the causal variant.</p>
<p>Traits can be either continuous, e.g. blood pressure, or discrete,
e.g. a disease. Basic analyses require information on the summarized
effect estimates (i.e. estimated regression coefficients) and their
corresponding standard errors for each snp in the genomic region and
each trait under consideration. These should be entered as numeric
matrices. If the traits are measured in non-independent studies
(i.e. those containing overlapping participants) analyses can be
adjusted to account for this. To do this three additional matrices are
required: (i) the pair-wise marginal correlations between the traits;
(ii) the pair-wise LD estimates between the snps in the region and;
(iii) the pair-wise estimates of the proportion of sample overlap
between study participants. We recommend reading our note (below and in
more detail our paper) on adjusting analyses to account for correlated
summary data and a-priori trait correlation before doing so in your
analyses.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>In the first part of this exercise we begin by loading the package
and some data needed to run analyses using HyPrColoc. For a given
region, a standard analysis requires data from two matrices, of equal
size, denoting: (i) a matrix of effect estimates (betas), with the
columns denoting the study traits and rows the snps, and; (ii) a matrix
of corresponding standard errors (ses).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jrs95/hyprcoloc" class="external-link">hyprcoloc</a></span><span class="op">)</span></span>
<span><span class="va">betas</span> <span class="op">&lt;-</span> <span class="fu">hyprcoloc</span><span class="fu">::</span><span class="va">test.betas</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">betas</span><span class="op">)</span></span>
<span><span class="co">#&gt;                      T1           T2            T3           T4           T5</span></span>
<span><span class="co">#&gt; rs6694014    0.02791630 -0.030353270 -0.0006508550 -0.015820079  0.029344113</span></span>
<span><span class="co">#&gt; rs11206477  -0.01333565 -0.007925434 -0.0220791782 -0.024533654 -0.006170044</span></span>
<span><span class="co">#&gt; rs978479     0.02789307 -0.031569213  0.0013910408 -0.016449156  0.030562284</span></span>
<span><span class="co">#&gt; rs6684892    0.01109516 -0.035811867 -0.0041582437 -0.001093336  0.021029899</span></span>
<span><span class="co">#&gt; rs149881092 -0.02058398  0.033028644  0.0732322501 -0.062564186  0.031519527</span></span>
<span><span class="co">#&gt; rs2081705    0.02804342 -0.031658194 -0.0006283532 -0.017953736  0.029719314</span></span>
<span><span class="co">#&gt;                      T6          T7          T8            T9         T10</span></span>
<span><span class="co">#&gt; rs6694014   -0.03739309 -0.05015619 -0.03963766 -0.0497666615 -0.01565694</span></span>
<span><span class="co">#&gt; rs11206477   0.08915364  0.08788543  0.07824464 -0.0364999498 -0.07140001</span></span>
<span><span class="co">#&gt; rs978479    -0.03667044 -0.04993080 -0.03999140 -0.0485026083 -0.01742117</span></span>
<span><span class="co">#&gt; rs6684892   -0.04342024 -0.05462565 -0.02610322 -0.0516383902 -0.01373675</span></span>
<span><span class="co">#&gt; rs149881092  0.03339484  0.08794777  0.06747322  0.0002877561 -0.01701753</span></span>
<span><span class="co">#&gt; rs2081705   -0.03798465 -0.05052849 -0.04150411 -0.0498085699 -0.01767236</span></span>
<span><span class="va">ses</span> <span class="op">&lt;-</span> <span class="fu">hyprcoloc</span><span class="fu">::</span><span class="va">test.ses</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ses</span><span class="op">)</span></span>
<span><span class="co">#&gt;                     T1         T2         T3         T4         T5         T6</span></span>
<span><span class="co">#&gt; rs6694014   0.01640805 0.01601499 0.01606328 0.01595009 0.01629387 0.01602093</span></span>
<span><span class="co">#&gt; rs11206477  0.01522992 0.01486595 0.01490668 0.01480196 0.01512465 0.01484631</span></span>
<span><span class="co">#&gt; rs978479    0.01642066 0.01602706 0.01607561 0.01596228 0.01630616 0.01603341</span></span>
<span><span class="co">#&gt; rs6684892   0.01738306 0.01696379 0.01701563 0.01689661 0.01726146 0.01696989</span></span>
<span><span class="co">#&gt; rs149881092 0.03917473 0.03823671 0.03833955 0.03807307 0.03890204 0.03825441</span></span>
<span><span class="co">#&gt; rs2081705   0.01643768 0.01604368 0.01609229 0.01597868 0.01632324 0.01604975</span></span>
<span><span class="co">#&gt;                     T7         T8         T9        T10</span></span>
<span><span class="co">#&gt; rs6694014   0.01618062 0.01625153 0.01639785 0.01626775</span></span>
<span><span class="co">#&gt; rs11206477  0.01499870 0.01506721 0.01522146 0.01508189</span></span>
<span><span class="co">#&gt; rs978479    0.01619313 0.01626394 0.01641083 0.01628007</span></span>
<span><span class="co">#&gt; rs6684892   0.01713951 0.01721824 0.01737041 0.01723253</span></span>
<span><span class="co">#&gt; rs149881092 0.03863522 0.03880163 0.03916327 0.03883610</span></span>
<span><span class="co">#&gt; rs2081705   0.01620976 0.01628044 0.01642748 0.01629694</span></span></code></pre></div>
<p>In the test dataset there are ten traits with summary association and
standard error data for around 1000 snps. The summary association data
are weakly correlated, owing to generating data from studies containing
overlapping participants (/samples). We can call the correlation and LD
matrices by typing</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trait.cor</span> <span class="op">&lt;-</span> <span class="fu">hyprcoloc</span><span class="fu">::</span><span class="va">test.corr</span></span>
<span><span class="va">ld.matrix</span> <span class="op">&lt;-</span> <span class="fu">hyprcoloc</span><span class="fu">::</span><span class="va">test.ld</span></span></code></pre></div>
<p>Note that in the sample data, traits 1-5 form a cluster of
colocalized traits; traits 6-8 form a distinct cluster of colocalized
traits and; traits 9-10 form distinct cluster of colocalized traits.</p>
<div class="section level3">
<h3 id="basic-set-up-assuming-independence-between-studies">Basic set-up: assuming independence between studies<a class="anchor" aria-label="anchor" href="#basic-set-up-assuming-independence-between-studies"></a>
</h3>
<p>By default HyPrColoc assumes that each trait is measured in a
distinct study, i.e. that the participants do not overlap between
studies, so that between study estimates of regression coefficients
(betas) are independent. The HyPrColoc function will automatically
employ the Bayesian divisive clustering algorithm to identify clusters
of colocalized traits. Each study (or trait) will be assigned a name
corresponding to their column position and similarly the snps will be
assigned names according to row position. To input trait and snp labels
we make use of the “trait.names” and “snp.id” variables, e.g. </p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"T"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">betas</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">rsid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">betas</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>, trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9164             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9018             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
<p>From the above output we see that, for each iteration of the
algorithm, HyPrColoc returns:</p>
<ol style="list-style-type: lower-roman">
<li>a cluster of putatively colocalized traits</li>
<li>the posterior probability that these traits are colocalized</li>
<li>the ‘regional association’ probability* (which is always &gt; the
posterior probability)</li>
<li>a candidate causal variant explaining the shared association</li>
<li>the proportion of the posterior probability explained by this
variant (which represents the HyPrColoc multi-trait fine-mapping
probability).</li>
</ol>
<p>*Note that a ‘large’ regional association probability is evidence
that one or more snps in the region have shared associations across the
traits, the result is similar to a PheWAS (Phenome-wide association
study) and we discuss this more later. The results above show that in
three iterations HyPrColoc identified that traits 1-5 form a cluster of
colocalized traits, traits 6-8 form a separate cluster of colocalized
traits and finally that traits 9 and 10 also colocalize. We see that the
cluster of traits 1-5 have a posterior probability of <span class="math inline">\(1\)</span> of being colocaized, hence the regional
association probability is also <span class="math inline">\(1\)</span>
and that the candidate snp rs11591147 explains 100% of the posterior.
Thus, there is very strong support that traits 1-5 colocalize and that
rs11591147 be taken as the candidate causal snp in the region. On the
other hand, traits 9 and 10 show strong evidence of colocalizing, having
a posterior probability of <span class="math inline">\(0.9\)</span>, but
there is weak evidence to support rs7524677 as the causal snp in the
region. This example helps to illustrate that: while a cluster of traits
can show strong evidence of colocalization, the snp most likely to
explain colocalization between the traits can be vague. We assess this
in more detail later, when considering credible sets of snps which
explain the colocalization signal (for each cluster of traits) c.f.
section “Computing a credible set of snps for each cluster of
colocalized traits”.</p>
<div class="section level4">
<h4 id="labelling-a-trait-as-either-continuous-or-binary-in-analyses">Labelling a trait as either continuous or binary in analyses<a class="anchor" aria-label="anchor" href="#labelling-a-trait-as-either-continuous-or-binary-in-analyses"></a>
</h4>
<p>For technical reasons, analyses have a dependence on whether a trait
is continuous or binary. To let HyPrColoc know which traits are
continuous (coded 0) or binary (coded 1) we use the “binary.outcome”
variable. For example, suppose the first three traits are binary, then
we type</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">binary.traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">betas</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">-</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>, trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, binary.outcomes <span class="op">=</span> <span class="va">binary.traits</span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9164             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9018             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="choosing-a-subset-of-traits-to-analyze">Choosing a subset of traits to analyze<a class="anchor" aria-label="anchor" href="#choosing-a-subset-of-traits-to-analyze"></a>
</h4>
<p>We can choose to assess evidence of colocalization across a subset of
the traits. For example, let’s suppose interest lies in assessing the
first two traits only, in this situation we make use of the
“trait.subset” variable, e.g.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>, trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, trait.subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T1"</span>, <span class="st">"T2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2              1             1    rs11591147</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                          1            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="computing-a-credible-set-of-snps-for-each-cluster-of-colocalized-traits">Computing a credible set of snps for each cluster of colocalized
traits<a class="anchor" aria-label="anchor" href="#computing-a-credible-set-of-snps-for-each-cluster-of-colocalized-traits"></a>
</h3>
<p>To output a fine-mapping score, i.e. a probability that a snp is
likley to be causal, for each of the snps in the region we make use of
the “snpscores” variable in analyses. When we do this HyPrColoc outputs
a ‘list’ of results where the first element contains the results
displayed previously and the second element containts the snp scores for
each of the clusters of colocalized traits identified. Let’s have a
look, if we now type</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>, trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, snpscores <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>then res is a list with two elements: “res[[1]]” and “res[[2]]”. The
element “res[[1]]” contains the results from before, i.e.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9164             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9018             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span></code></pre></div>
<p>In the present example, “res[[2]]” is a list of length 3 containing
the snp scores for each of the 3 clusters of colocalized traits
(presented in the same order as the clusters in “res[[1]]”), i.e
“res[[2]][[1]]” is a vector and each element of the vector denotes the
probability that a given snp is the causal variant for the cluster of
traits 1-5. “res[[2]][[2]]” returns the same information but now for the
cluster of putatively colocalized traits 6-8 and “res[[2]][[3]]” for the
cluster 9-10. For example, the first few snp scores for cluster 1
(traits 1-5) are:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;    rs6694014   rs11206477     rs978479    rs6684892  rs149881092    rs2081705 </span></span>
<span><span class="co">#&gt; 4.109458e-67 3.004278e-68 5.637121e-67 6.486572e-68 6.063410e-66 5.765875e-67</span></span></code></pre></div>
<p>HyPrColoc has an inbuilt function to triage this information down to
a “credible set” of snps for each cluster which uses the output from the
“hyprcoloc” function when “snpsocres = TRUE”. By default, the function
identifies a credible set of snps which explains 95% of the posterior
probability of colocalization for each cluster of colocalized traits. We
this using the “cred.sets” function, i.e.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/cred.sets.html">cred.sets</a></span><span class="op">(</span><span class="va">res</span>, value <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; rs11591147 </span></span>
<span><span class="co">#&gt;          1 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt;  rs12117612   rs7532349  rs11206481  rs12145624  rs12724445  rs12126037 </span></span>
<span><span class="co">#&gt; 0.419677651 0.419677651 0.025324539 0.022969148 0.022969148 0.022332481 </span></span>
<span><span class="co">#&gt;  rs13375783   rs1544909 </span></span>
<span><span class="co">#&gt; 0.014513746 0.009991906 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt;  rs7524677  rs7547776  rs7524783  rs6701789  rs1035817  rs7530321  rs7520033 </span></span>
<span><span class="co">#&gt; 0.07626197 0.07626197 0.07626197 0.07626197 0.07626197 0.05349125 0.05349125 </span></span>
<span><span class="co">#&gt;  rs7553410 rs11206486  rs7524899  rs6681189  rs7537797 rs11206485 rs10888890 </span></span>
<span><span class="co">#&gt; 0.05349125 0.02231383 0.02231383 0.02231383 0.02231383 0.02098626 0.02098626 </span></span>
<span><span class="co">#&gt;  rs2114574  rs6664660  rs6588539 rs11206489  rs6668051 rs10218512 rs11206492 </span></span>
<span><span class="co">#&gt; 0.02098626 0.02098626 0.02098626 0.02098626 0.02098626 0.02098626 0.02098626 </span></span>
<span><span class="co">#&gt;  rs4634950  rs7539163  rs7529244  rs6687414  rs7538808 rs11206487 rs12745865 </span></span>
<span><span class="co">#&gt; 0.02098626 0.02098626 0.02098626 0.02098626 0.02098626 0.01203687 0.01203687 </span></span>
<span><span class="co">#&gt; rs12725873 </span></span>
<span><span class="co">#&gt; 0.01203687</span></span></code></pre></div>
<p>We see that snp “rs11591147” explains ‘all’ of the posterior
probability of colocalization between the cluster of traits 1-5. This is
therefore a strong candidate causal variant in the region. However, in
the second cluster of traits (6-8), there is no clear candidate causal
variante as both “rs12117612” and “rs7532349” explain equal amounts of
the posterior probability of colocalization and, as we now show, are in
perfect LD with one another:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld.matrix</span><span class="op">[</span><span class="st">"rs12117612"</span>, <span class="st">"rs7532349"</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 1</span></span></code></pre></div>
<p>Together these two variants explain nearly 90% of the posterior
probability of colocalization and it is therefore likely* that the
traits do colocalize, however it is impossible (using HyPrColoc) to
prioritise one of these variants over the other. To prioritise one of
these snps external information would need to be integrated, which can
be (e.g.) biological information or by adding another layer to the
analysis in which the same trait, or traits, are measured in different
studies.</p>
<p>*One natural question to ask is: if the two snps are in perfect LD
why have we deduced that the traits colocalize? The result is a
consequence of our choice of the prior probability of colocalization.
The number of causal configurations which reflect colocalization between
the traits is much (much!) smaller than configurations which do not
represent colocalization between the traits. This means that if we were
to set all prior configuration probabilities to be equal we would almost
certainly never identify colocalization between the traits, even if the
shared associations between traits were truly owing to a colocalization
mechanism. Thus, a colocalization causal confiugration prior must be
larger than a non-colocalization causal configuration prior for general
(discovery) analyses. But how much larger? There is no simple answer to
this. It is complicated and we therefore dedicate more time to discuss
this now.</p>
</div>
</div>
<div class="section level2">
<h2 id="an-analysis-protocol-assessing-stability-of-clusters-via-a-sensitivity-analysis">An analysis protocol: assessing stability of clusters via a
sensitivity analysis<a class="anchor" aria-label="anchor" href="#an-analysis-protocol-assessing-stability-of-clusters-via-a-sensitivity-analysis"></a>
</h2>
<p>It is important that users of the software ackowledge that the
performance of HyPrColoc is particularly dependendant on the choice of
prior configuration probabilities used (and any associated
hyper-parameters) as well as the choice of regional and alignment
thresholds, as these combine to quantify a lower bound with which we
accept that a cluster of traits colocalize (i.e. clusters are identified
when <span class="math inline">\(P_RP_{A}\geq
P^{\ast}_{R}P^{\ast}_{A}\)</span>). In some situations this senstivity
might be modest, whilst in others it might be large. For example,
through extensive simulations, we note that in the analysis of large
numbers of traits, using only the default algorithm settings, can
regularly result in the trait clusters containing (typically only a
single) false positive. Avoiding this issue is complex as it is unlikely
that there exists a one-size-fits-all approach to setting the prior
configuration probabilities and likewise the regional and alignment
threshold parameters. Hence, to go someway to addressing these issues we
provide an analysis protocol template, to assess the strength of any
conclusions.</p>
<p>At the end of this section we introduce a function “sensitivity.plot”
which, on varying the input values for the regional and alignment
thresholds as well as the prior probability of colocalization, returns a
heat-map that helps us to visuale how stable the clusters of colocalized
traits are to variations in the algorithms input parameters. We expect
this function to be a part of standard analyses, helping users to
pinpoint the best candidate clusters of colocalized traits for follow-up
analyses.</p>
<div class="section level3">
<h3 id="assessing-sensitivity-to-changes-in-the-prior-configuration-parameters">Assessing sensitivity to changes in the prior configuration
parameters<a class="anchor" aria-label="anchor" href="#assessing-sensitivity-to-changes-in-the-prior-configuration-parameters"></a>
</h3>
<p>An important feature of the HyPrColoc software is that it allows for
some sensitivity to the choice of causal configuration priors to be
assessed. Two prior choices are presented: (i) conditionally uniform
priors, which assumes that all causal configurations relating to a given
hypotheses are equally likely, and; (ii) variant specific prioris,
which, for each variant, focuses on tuning the probability that a
variant is colocalized with a subset of traits, for all possible
subsets. HyPrColoc has some dependency to both the type of prior
configuration assumed and the parameters associated with each prior
set-up. In this section we aim to outline an approach to assessing any
sensitivity to these choices and moreover some guidance on drawing
conclusions from sesnsitivity analyses.</p>
<p>HyPrColoc jointly assesses colocalization across multiple traits and
we use this detail to our advantage in our sensitivity analyses. The
general principle is as follows: perform two or more analyses in which
the probability that a variant is associated with more than one trait
decreases and for each assessment compute the number of traits within
each cluster that overlapp between analyses. Any overlap in the
collection of traits within each cluster, which use different choices of
prior paramater values, would then be a ‘stable’ cluster of colocalized
traits. We might wish to do this for each collection of putatively
colocalized traits identified from the primary analyses (which may or
may not use the default settings) or repeat the whole analysis again
under a different parameter specification.</p>
<div class="section level5">
<h5 id="conditionally-uniform-configuration-priors">Conditionally uniform configuration priors:<a class="anchor" aria-label="anchor" href="#conditionally-uniform-configuration-priors"></a>
</h5>
<p>The conditionally uniform prior assigns all non-null hypotheses
(i.e. those in which at least one trait has a causal variant in the
region) the same probaility. Note that this does not mean that all
causal configuration prior probailities are the same, rather that the
prior probabilities of each causal configuration associated with a
distinct hypothesis must add up to the same amount. This means that the
conditionally uniform prior benefits from automatically adjusting to
both the number of traits and the number of snps included in analyses.
The approach requires specificying one parameter only, “prior.1”, which
is the ratio of the prior probability of a hypothesis in which at least
one trait has a causal variant in the region <span class="math inline">\(P(H_{j})\)</span> divided by the null-hypothesis
that no traits have a causal variant in the region <span class="math inline">\(P(H_{0})\)</span>, i.e. <span class="math inline">\(prior.1 = \frac{P(H_{j})}{P(H_{0})}\)</span>. By
default, we set <span class="math inline">\(prior.1=10^{-4}\)</span>, so
that a non-null hypothesis is 10000 times less likely a-priori relative
to the null hypothesis. The default choice <span class="math inline">\(10^{-4}\)</span> is motivated in part by the role
of <span class="math inline">\(p1\)</span> in the widely used sofwtare
“COLOC”, where <span class="math inline">\(p1\)</span> is the prior
prbability that a snp is causally associated with one trait in the
region whereas <span class="math inline">\(prior.1\)</span> can be
thought of as the prior probability of a hypothesis, in which at least
one trait has a causal variant in the region. As the number of traits in
the sample increases the default choice of <span class="math inline">\(prior.1\)</span> may be become increasingly
inappropriate. It is therefore important to explore sensitivity of any
results to the specification of “prior.1” which we now do.</p>
<p>As the number of traits in the test dataset is ‘small’, we choose to
repeatedly run HyPrColoc and for each run change the value <span class="math inline">\(prior.1\in {10^{-4}, 10^{-10}, 10^{-20}, 10^{-25},
10^{-100}}\)</span>. If the number of traits in the sample were large we
might consider selecting a cluster of traits with which to assess
cluster stability.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prior.options</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e-4</span>, <span class="fl">1e-10</span>, <span class="fl">1e-20</span>, <span class="fl">1e-25</span>, <span class="fl">1e-100</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">prior.options</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>,</span>
<span>    trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>,</span>
<span>    uniform.priors <span class="op">=</span> <span class="cn">TRUE</span>, prior.1 <span class="op">=</span> <span class="va">i</span>, reg.steps <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"prior.1 = "</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; [1] "prior.1 = 1e-04"</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9888             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9799             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span>
<span><span class="co">#&gt; [1] "prior.1 = 1e-10"</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9888             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9799             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span>
<span><span class="co">#&gt; [1] "prior.1 = 1e-20"</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000        1.0000    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9887        1.0000    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9797        0.9998     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span>
<span><span class="co">#&gt; [1] "prior.1 = 1e-25"</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5              1         1.000    rs11591147</span></span>
<span><span class="co">#&gt; 2         2               None             NA         0.000          &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 3         3               None             NA         0.046          &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 4         4               None             NA         0.000          &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 5         5               None             NA         0.000          &lt;NA&gt;</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                          1          &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2                         NA            T7</span></span>
<span><span class="co">#&gt; 3                         NA           T10</span></span>
<span><span class="co">#&gt; 4                         NA            T8</span></span>
<span><span class="co">#&gt; 5                         NA            T9</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span>
<span><span class="co">#&gt; [1] "prior.1 = 1e-100"</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1   None             NA             0            NA</span></span>
<span><span class="co">#&gt; 2         2   None             NA             0            NA</span></span>
<span><span class="co">#&gt; 3         3   None             NA             0            NA</span></span>
<span><span class="co">#&gt; 4         4   None             NA             0            NA</span></span>
<span><span class="co">#&gt; 5         5   None             NA             0            NA</span></span>
<span><span class="co">#&gt; 6         6   None             NA             0            NA</span></span>
<span><span class="co">#&gt; 7         7   None             NA             0            NA</span></span>
<span><span class="co">#&gt; 8         8   None             NA             0            NA</span></span>
<span><span class="co">#&gt; 9         9   None             NA             0            NA</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                         NA            T1</span></span>
<span><span class="co">#&gt; 2                         NA            T4</span></span>
<span><span class="co">#&gt; 3                         NA            T7</span></span>
<span><span class="co">#&gt; 4                         NA           T10</span></span>
<span><span class="co">#&gt; 5                         NA            T2</span></span>
<span><span class="co">#&gt; 6                         NA            T8</span></span>
<span><span class="co">#&gt; 7                         NA            T3</span></span>
<span><span class="co">#&gt; 8                         NA            T9</span></span>
<span><span class="co">#&gt; 9                         NA            T5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
<p>The results reveal that: in the sample dataset provided and when
using the conditionally uniform priors, the results from HyPrColoc are
weakly dependendent on the specification of “prior.1”. In particular,
across the range <span class="math inline">\(prior.1 \in [10^{-4},
10^{-20}]\)</span> the HyPrColoc results do not change beyond the fourth
decimal place. We might conclude therefore that the three clusters of
traits 1-5; 6-8 and separately 9-10 are stable clusters of colocalized
traits. When <span class="math inline">\(prior.1 = 10^{-25}\)</span>
only the cluster of traits <span class="math inline">\(1-5\)</span> are
deemed to colocalize and when <span class="math inline">\(prior.1 =
10^{-100}\)</span> no traits are deemed to colocalize. So, do none of
the traits truly colocalize? No, it should be noted that we specified
the values <span class="math inline">\(prior.1\in {10^{-4}, 10^{-10},
10^{-20}, 10^{-25}, 10^{-100}}\)</span> not because these are all
‘reasonable’ values to assess prior sensitivty, but rather that in the
test dataset we tried to find a value, i.e. <span class="math inline">\(prior.1=10^{-25}\)</span>, below which evidence
supporting the traits forming clusters of colocalized traits is ‘small’
so that we fail to detect colocalization between the traits. In practice
and in the absence of a prior belief concerning <span class="math inline">\(prior.1\)</span>, a reasonable choice will depend
on the number of traits under consideration. However, to guide analyses
we suggest comparing results using the default <span class="math inline">\(10^{-4}\)</span> with those when <span class="math inline">\(prior.1 = 10^{-5}\)</span>, i.e. an order of
magnitude reduction in “prior.c”, for sensitivity analyses.</p>
<p>Note that, using the conditionally uniform prior with default <span class="math inline">\(prior.1=10^{-4}\)</span>, the posterior
probabilities computed for each cluster of colocalized traits is larger
than those computed using the default parameters values under the
variant specific prior configuration set-up; we now illustrate this.</p>
</div>
<div class="section level5">
<h5 id="variant-specific-configuration-priors-default">Variant specific configuration priors (default):<a class="anchor" aria-label="anchor" href="#variant-specific-configuration-priors-default"></a>
</h5>
<p>The variant specific causal configuration priors in HyPrColoc are, in
some sense, a multi-trait extention to the causal configuration priors
used in COLOC (for two traits they are identical). Users familiar with
the COLOC software will no doubt know that the performance of COLOC is
sensitive to the choice of the causal configuration prior parameters.
The HyPrColoc multi-trait extention of this prior set-up is also
sensitive to the choice of prior parameters.</p>
<p>The HyPrColoc variant specific priors require the specification of
two parameters: “prior.1” and “prior.c”. The parameter “prior.1” denotes
the probability that a snp is associated with a single trait. “prior.c”
is the conditional colocalization prior. We write <span class="math inline">\(prior.c = 1-prior.2\)</span>, so that: (i) <span class="math inline">\(1-prior.2\)</span> is the prior probaility that a
snp is associated with an additonal trait given that it is associated
with one trait; (ii) <span class="math inline">\(1-(prior.2)^2\)</span>
is the prior probability that the snp is associated with a third trait
given it is associated with two other traits; (iii) <span class="math inline">\(1-(prior.2)^3\)</span> is the prior probability
that the snp is associated with a fourth trait given three and so on… By
default <span class="math inline">\(prior.1 = 10^{-4}\)</span>, hence
the prior probability that any snp is a ssociated with a single trait is
1 in 10000 (matching that used in COLOC). The default for <span class="math inline">\(prior.c = 0.02\)</span>, meaning that the prior
probability that the snp is associated with a second trait, conditional
on association with one trait, is 1 in 50; the prior probaility that it
is associated with a third trait given association with two other traits
is around 1 in 25; and so on… Notice that the marginal probability that
a snp is associated with an increasing number of traits is always small
and monotonic decreasing, e.g. <span class="math inline">\(\frac{1}{10^{4}*50*25*\dots}\)</span>.</p>
<p>When using the variant specific prior, results tend to be most
sensitive to the choice of <span class="math inline">\(prior.c\)</span>
as this controls the prior probability of each additional colocalized
trait. Hence, we focus our sensitivity analysis on varying this
parameter. We consider the range of values <span class="math inline">\(prior.c \in\{0.05, 0.02, 0.01, 0.005\}\)</span>,
meaning that the probability of a snp being associated with a second
trait given its association with one trait can be: (i) 1 in 20,
i.e. <span class="math inline">\(prior.c=0.05\)</span>; (ii) 1 in 50,
i.e. <span class="math inline">\(prior.c=0.02\)</span>; (iii) 1 in 100,
i.e. <span class="math inline">\(prior.c=0.01\)</span>; (ii) 1 in 200,
i.e. <span class="math inline">\(prior.c=0.005\)</span>,</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prior.1</span> <span class="op">&lt;-</span> <span class="fl">1e-4</span></span>
<span><span class="va">prior.c.options</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.02</span>, <span class="fl">0.01</span>, <span class="fl">0.005</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">prior.c.options</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>,</span>
<span>    trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>,</span>
<span>    uniform.priors <span class="op">=</span> <span class="cn">FALSE</span>, prior.1 <span class="op">=</span> <span class="va">prior.1</span>, prior.c <span class="op">=</span> <span class="va">i</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"prior.1 = "</span>, <span class="va">prior.1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"prior.c = "</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; [1] "prior.1 = 1e-04" "prior.c = 0.05" </span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9643             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9582             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span>
<span><span class="co">#&gt; [1] "prior.1 = 1e-04" "prior.c = 0.02" </span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9164             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9018             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span>
<span><span class="co">#&gt; [1] "prior.1 = 1e-04" "prior.c = 0.01" </span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.8464             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.8211             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span>
<span><span class="co">#&gt; [1] "prior.1 = 1e-04" "prior.c = 0.005"</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.7342             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.6965             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
<p>The results indicate that traits 1-5 appear to form a ‘stable’
cluster of colocalized traits, i.e. across the range of prior
probabilities considered the traits 1-5 always colocalize. For the two
separate clusters comprising traits 6-8 and traits 9-10, the results
differ depending on the choice of <span class="math inline">\(prior.c\)</span>. Firstly, we note that the
posterior probability of colocalization for each of these clusters
decreases as a function of <span class="math inline">\(prior.c\)</span>,
but the reduction is not the same between clusters because the cluster
sizes differ. That is, for <span class="math inline">\(prior.c
\in\{0.05, 0.02, 0.01\}\)</span> the posterior probability of the
cluster of size three (traits 6-8) decreases from <span class="math inline">\(0.964\)</span> to <span class="math inline">\(0.846\)</span>, whereas for the cluster of size
two (traits 9-10) it reduces from <span class="math inline">\(0.958\)</span> to <span class="math inline">\(0.821\)</span>. In summary, the larger the number
of traits in a cluster the less sensitive results are to the
specification of <span class="math inline">\(prior.c\)</span>. We see
this as a positive feature of the HyPrColoc variant specific prior.
However, if <span class="math inline">\(prior.c\)</span> is reduced by
an order of magnitude from <span class="math inline">\(10^{-2}\)</span>
to <span class="math inline">\(10^{-3}\)</span>, i.e. a drop from 1 in
100 to 1 in 1000, we note that of the two clusters of traits (6-8 and
9-10) only traits 7-8 appear to colocalize and the posterior probability
of colocalization is weaker than before <span class="math inline">\(0.739\)</span>. In the absense of a prior belief
concerning <span class="math inline">\(prior.c\)</span>, a reasonable
value for <span class="math inline">\(prior.c\)</span> will depend on
the number of traits under consideration. As a guide, we suggested
assessing results from <span class="math inline">\(prior.c\in \{0.02,
0.01\}\)</span>, i.e. the HyPrColoc default and when considering that
the probability of a snp having an additional association given an
association with one trait is 1 in 100 (our “sensitivity.plot” function
will do this for you).</p>
<p>One additional point to consider is evidential strength, if we look
at the output above when setting <span class="math inline">\(prior.c\in
\{0.02, 0.01\}\)</span>, we note that the posterior probability of
colocalization between the traits in the clusters 6-8 and 9-10 both fall
below <span class="math inline">\(0.85\)</span>. Thus, setting a more
stringent threshold for colocalization between the traits, e.g. only
return results when a cluster of traits colocalize with posterior
probability above <span class="math inline">\(0.85\)</span>, will
necessarily alter the results presented by HyPrColoc; which we now
demonstrate.</p>
</div>
<div class="section level5">
<h5 id="evidential-strength-and-the-regional-and-alignment-thresholds">Evidential strength and the regional and alignment thresholds<a class="anchor" aria-label="anchor" href="#evidential-strength-and-the-regional-and-alignment-thresholds"></a>
</h5>
<p>HyPrColoc estimates the posterior probability of colocalization
between a cluster of traits by multiplying a regional association
probability <span class="math inline">\(P_{R}\)</span> and an alignment
probability <span class="math inline">\(P_{A}\)</span>, i.e. <span class="math inline">\(P_{R}*P_{A}\)</span>. The algorithm concludes that
a cluster of traits colocalize if these values are above two threshold
values: the regional probability threshold <span class="math inline">\(P_{R}^{\ast}\)</span> and the alignment
probability threshold <span class="math inline">\(P_{A}^{\ast}\)</span>.
As noted from the previous section, the posterior probability can vary
by the choice of configuration priors used. Consequently, the default
regional and alignment thresholds vary when using either the
conditionally uniform configuration priors, i.e. <span class="math inline">\(P_R^{\ast}=P_{A}^{\ast}=0.7\)</span>, or the
variant specific priors, i.e. <span class="math inline">\(P_R^{\ast}=P_{A}^{\ast}=0.5\)</span>. Accordingly,
the algorithm will deduce that a set of traits are colocalized when
<span class="math inline">\(P_RP_{A}\geq 0.49\)</span> using CU priors
and <span class="math inline">\(P_RP_{A}\geq 0.25\)</span> when using VS
priors. Note, these parameter choices are a consequence of extensive
testing in simulation scenarios, aiming to maximise the true detection
rate whilst minimising the number of false positives. We therefore DO
NOT recommend reducing the value of either of these parameters. The
defaults should be viewed as lower bounds.</p>
<p>If interest lies in identifying sets of colocalized traits above a
certain posterior probability, more stringent regional and alignment
threshold parameters can be chosen. For example if we set <span class="math inline">\(P_R^{\ast}=P_{A}^{\ast}=0.95\)</span> so that
colocalized traits are identified when <span class="math inline">\(P_RP_{A}\geq 0.9025\)</span> we see that</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>,</span>
<span>  trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, uniform.priors <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reg.thresh <span class="op">=</span> <span class="fl">0.95</span>, align.thresh <span class="op">=</span> <span class="fl">0.95</span></span>
<span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000        1.0000    rs11591147</span></span>
<span><span class="co">#&gt; 2         2             T7, T8         0.9827        1.0000    rs12117612</span></span>
<span><span class="co">#&gt; 3         3               None             NA        1.0000          &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 4         4               None             NA        0.9653          &lt;NA&gt;</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                      1.000          &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2                      0.428          &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 3                         NA           T10</span></span>
<span><span class="co">#&gt; 4                         NA            T9</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
<p>Hence, again we identify that the clusters of traits with very strong
evidence of colocalizing are traits 1-5 and separately 7-8. However,
some of you might notice that in our previous assessment of the data
traits 6-8 colocalized with posterior probability <span class="math inline">\(0.916\)</span> and traits 9-10 with posterior
probability <span class="math inline">\(0.902\)</span>. Traits 9-10 fall
below our posterior threshold, however traits 6-8 do not. Why is when we
set <span class="math inline">\(P_R^{\ast}=P_{A}^{\ast}=0.95\)</span> do
we not identify the full cluster 6-8 rather than just 7-8? The answer is
that while the posterior probability of traits 6-8 colocalizing is <span class="math inline">\(0.916\)</span> either the regional or alignment
probability is not larger than the threshold <span class="math inline">\(0.95\)</span> (when trait 6 is included in the
cluster). In this situation the algorithm will fail to identify the
maximum number of clusters of traits which colocalize above <span class="math inline">\(0.902\)</span>.</p>
<p>In our extensive testing of the HyPrColoc algorithm we noticed that
we can avoid this if we instead set both the regional and alignment
statistics to the minimum posterior probability that determines that a
cluster of traits are colocalized. Hence, in the above example we would
instead set <span class="math inline">\(P_R^{\ast}=P_{A}^{\ast}=0.9025\)</span>, i.e. </p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>,</span>
<span>  trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, uniform.priors <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reg.thresh <span class="op">=</span> <span class="fl">0.9025</span>, align.thresh <span class="op">=</span> <span class="fl">0.9025</span></span>
<span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9164             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3               None             NA             1          &lt;NA&gt;</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000          &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2                     0.4197          &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 3                         NA           T10</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
<p>which returns the full cluster of traits 6-8.</p>
<p>This results explains why the default values for <span class="math inline">\(P_R^{\ast}\)</span> and <span class="math inline">\(P_{A}^{\ast}\)</span> are set to be ‘equal’ and
‘both’ reflect the minimum probability with which we accept
clocalization, not the product <span class="math inline">\(P_{R}*P_{A}\)</span>.</p>
</div>
<div class="section level5">
<h5 id="analysis-protocol-a-one-stop-sensitivity-assessment">Analysis protocol: a one stop sensitivity assessment<a class="anchor" aria-label="anchor" href="#analysis-protocol-a-one-stop-sensitivity-assessment"></a>
</h5>
<p>As discussed above, results from a colocalization analyses might be
sensitive to the choice of: (i) algorithm thresholds,
i.e. colocalization cut-off <span class="math inline">\(=P^{\ast}_{R}*P^{\ast}_{A}\)</span>, and; (ii) the
prior belief about traits sharing a causal variant. It is therefore
essential that we explore any sensitivity, particularly if we are aiming
to identify reliable candidate clusters of colocalized traits. HyPrColoc
allows you to do this through a single function “sensitivity.plot”. The
function repeatedly calls “hyprcoloc” for various user defined choices
of the regional and alignment thresholds as well as the colocalization
prior (“prior.c”). The default values for these parameters start from
the algorithms default settings and then increase these so that
identification of clusters of colocalized traits becomes more
challenging (i.e. traits must have stronger evidence of colocalization
in order to be detected). For example</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sensitivity.plot.html">sensitivity.plot</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>,</span>
<span>  trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>,</span>
<span>  reg.thresh <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.7</span>, <span class="fl">0.8</span>, <span class="fl">0.9</span><span class="op">)</span>, align.thresh <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.7</span>, <span class="fl">0.8</span>, <span class="fl">0.9</span><span class="op">)</span>, prior.c <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.02</span>, <span class="fl">0.01</span>, <span class="fl">0.005</span><span class="op">)</span>, equal.thresholds <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="hyprcoloc_files/figure-html/unnamed-chunk-17-1.png" width="480"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Or by fixing the regional and aligment thresholds to be equal</span></span>
<span><span class="co"># sensitivity.plot(betas, ses, trait.names = traits, snp.id=rsid,</span></span>
<span><span class="co">#                  reg.thresh = c(0.6,0.7,0.8,0.9), prior.c = c(0.02, 0.01, 0.005), equal.thresholds = TRUE);</span></span></code></pre></div>
<p>The output is a heat-map which helps us to visualise the clusters of
colocalised traits. Traits which cluster across all values considered
are given a score of 1 and traits which never cluster are given a score
of 0; traits which ocassionally cluster have a score between 0 and 1.
Hence, our confidence in a cluster increases with increasing score -
appearing as a darker block of clustered traits in the heat-map. The
heat-map reflects what we have already seen from our individual
analyses: traits 1-5 form a very strong cluster of colocalized traits;
traits 6-8 form a second cluster, however for certain values of the
input parameters trait 6 is dropped from this cluster and; traits 9-10
form a thrid cluster, however identification of this cluster is
difficult for more stringent values of the thresholds and the prior
probability of colocalization. It is clear from our figure that there
are three distinct clusters of colocalized traits in our sample,
matching the true data generating mechanism. However, our most confident
clusters of colocalised traits are traits 1-5 and separately traits 7-8.
Although we have set the benchmark quite high, the two clusters 6-8 and
9-10 are present in around 70% of the (sensitivity) scenarios considered
making them reasonable candidates also. To see this we can require that
“sensitivity.plot” returns the similarity matrix used to plot the
heat-map:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sensitivity.plot.html">sensitivity.plot</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>,</span>
<span>  trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>,</span>
<span>  reg.thresh <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.7</span>, <span class="fl">0.8</span>, <span class="fl">0.9</span><span class="op">)</span>, align.thresh <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.7</span>, <span class="fl">0.8</span>, <span class="fl">0.9</span><span class="op">)</span>, prior.c <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.02</span>, <span class="fl">0.01</span>, <span class="fl">0.005</span><span class="op">)</span>, equal.thresholds <span class="op">=</span> <span class="cn">FALSE</span>, similarity.matrix <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="hyprcoloc_files/figure-html/unnamed-chunk-18-1.png" width="480"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># heatmap is found by typing</span></span>
<span><span class="va">heatmap.plot</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># (plotted previously)</span></span>
<span><span class="co"># similarity matrix is found by typing</span></span>
<span><span class="va">sim.mat</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">sim.mat</span></span>
<span><span class="co">#&gt;     T1 T2 T3 T4 T5   T6   T7   T8        T9       T10</span></span>
<span><span class="co">#&gt; T1   1  1  1  1  1 0.00 0.00 0.00 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; T2   1  1  1  1  1 0.00 0.00 0.00 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; T3   1  1  1  1  1 0.00 0.00 0.00 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; T4   1  1  1  1  1 0.00 0.00 0.00 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; T5   1  1  1  1  1 0.00 0.00 0.00 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; T6   0  0  0  0  0 1.00 0.75 0.75 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; T7   0  0  0  0  0 0.75 1.00 1.00 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; T8   0  0  0  0  0 0.75 1.00 1.00 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; T9   0  0  0  0  0 0.00 0.00 0.00 1.0000000 0.6666667</span></span>
<span><span class="co">#&gt; T10  0  0  0  0  0 0.00 0.00 0.00 0.6666667 1.0000000</span></span></code></pre></div>
<p>We se that in 75% of scenarios traits 6-8 form a cluserter of
colocalized traits and in around 67% of scenarios traits 9-10 form a
cluster of colocalized traits.</p>
</div>
</div>
<div class="section level3">
<h3 id="the-bayesian-divisive-clustering-algorithm">The Bayesian divisive clustering algorithm<a class="anchor" aria-label="anchor" href="#the-bayesian-divisive-clustering-algorithm"></a>
</h3>
<p>The Bayesian divisive clustering algorithm identifies clusters of
colocalized traits in the sample. The algorithm is automatically used
when all traits are identified as not sharing a causal variant. For each
iteration of the algorithm, the goal is to identify a subset of traits
with the greatest evidence of colocalization. HyPrColoc does this using
one of two (user defined) approaches: (i) the regional or (ii) alignment
selection. The regional selection criterion is computed from a
collection of hypotheses which assume that all traits do not colocalize
because one of the traits does not have a causal variant in the region.
The alignment selection criterion, however, is computed from hypotheses
which assume that all traits do not colocalize because one of the traits
has a causal variant elsewhere in the region. The alignment selection
process searches a much larger amount of the causal configuration space
and is thus more computationally expensive (but potentially more robust)
than regional selection. Typically, both strategies perform similarly
and we therefore chose the regional selection as the default
setting.</p>
<div class="section level4">
<h4 id="assessing-differences-between-the-bayesian-divisive-clustering-criteria">Assessing differences between the Bayesian divisive clustering
criteria<a class="anchor" aria-label="anchor" href="#assessing-differences-between-the-bayesian-divisive-clustering-criteria"></a>
</h4>
<p>The regional selection criterion:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>, trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, bb.selection <span class="op">=</span> <span class="st">"regional"</span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9164             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9018             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
<p>The alignment selection criterion:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>, trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, bb.selection <span class="op">=</span> <span class="st">"align"</span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9164             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9018             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.4197            NA</span></span>
<span><span class="co">#&gt; 3                     0.0763            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
<p>Hence, in our test dataset there is no difference in the results when
using either the reginoal or the alignment selection criterion.</p>
</div>
<div class="section level4">
<h4 id="switching-the-algorithm-off">Switching the algorithm off<a class="anchor" aria-label="anchor" href="#switching-the-algorithm-off"></a>
</h4>
<p>We can choose to switch the Bayesian divisive clustering algorithm
off and assess whether all traits colocalize.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>, trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, bb.alg <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;                                    traits posterior_prob regional_prob</span></span>
<span><span class="co">#&gt; 1 T1, T2, T3, T4, T5, T6, T7, T8, T9, T10              0        0.0132</span></span>
<span><span class="co">#&gt;   candidate_snp posterior_explained_by_snp</span></span>
<span><span class="co">#&gt; 1    rs11591147                          1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="mapping-pleiotropy-an-alternative-to-phewas">Mapping pleiotropy: an alternative to PheWAS<a class="anchor" aria-label="anchor" href="#mapping-pleiotropy-an-alternative-to-phewas"></a>
</h2>
<p>The regional association can be used as an alternative to a PheWAS
assessment. To see this, we first discuss how the regional association
probability is computed.</p>
<p>For each of <span class="math inline">\(Q\)</span> snps in the
region, we compute the probability that a snp is associated with all
traits and divide this by the sum of the probabilities that: (i) one of
the <span class="math inline">\(Q\)</span> snps is associated with all
traits; (ii) one of the snps is associated with all but one trait; (iii)
one of the snps is associated with all but two traits and so on… Notably
there is no restriction on the number of causal variants in the region
under a regional association assessment. A ‘large’ (close to 1) regional
association probability indicates that one or more snps in the region
share an association with all traits. This may be owing to one or more
shared causal variants between the traits or because all traits have one
or more causal variants and at least one causal variant per trait is in
strong LD with a causal variant from each of the other traits. Small
values indicate that there is not a snp in the region associated with
all traits, this does not preclude the possibly that one or more snps
might be associated with a subset of the traits, however. The divisive
clustering algorithm contains the option (bb.selection = “reg.only”) to
identify clusters of traits which share a regional association only,
avoiding the alignment phase and therefore the colocalization
assessment. This function therefore allows us to assess evidence of
plieotropy between subsets (i.e. clusters) of traits.</p>
<p>Clusters of regionally associated traits are identified when the
regional association probability is above a user defined threshold
(default <span class="math inline">\(P_{R}^{\ast}= 0.5\)</span>, which
is likely to be anti-conservative and therefore should be increased). As
an example, below we assess any evidence of regional associations
between the traits in the sample dataset and we set the threshold
regional association probability to be <span class="math inline">\(0.9\)</span>:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>, trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, bb.selection <span class="op">=</span> <span class="st">"reg.only"</span>, reg.thresh <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ptm</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;    0.05    0.00    0.05</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration              traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1  T1, T2, T3, T4, T5              0        1.0000    rs11591147</span></span>
<span><span class="co">#&gt; 2         2 T6, T7, T8, T9, T10              0        0.9473    rs12117612</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.3137            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
<p>The results indicate that there exists one or more snps that share
associations with traits 1-5, a result anticipated from the results of
our previous colocalization analysis. However, by avoiding a
colocalization assessment we identify that traits 6-10 all appear to
have a shared association with at least one snp, the most likely shared
snp is rs12117612 which explains 31.4% of the regional association
probability. This contrasts with the results from our earlier
colocalization assessment which indicated that traits 6-10 do not all
colocalize and highlights some distinction between traits which have a
shared genetic associations (PheWAS) and those which colocalize at a
single causal variant.</p>
<p>As a final point of note, the regional association statistic is
computed from a much smaller number of causal configurations than a full
colocalization assessment. The upshot is that the Bayesian divisive
clustering algorithm, using only the regional statistic, can rapidly
identify clusters of traits which shared genetic associations amongst
very large numbers of traits. As an example, try employing a regional
only assessment to the dataset of 1000 traits that we build later in the
vignette. We can also assess any sensitivity in our results to the
choice of regional threshold using the “sensitivity.plot” function,
which we introduce in the next section.</p>
</div>
<div class="section level2">
<h2 id="analysing-large-numbers-of-traits">Analysing large numbers of traits<a class="anchor" aria-label="anchor" href="#analysing-large-numbers-of-traits"></a>
</h2>
<p>A common question we are asked is: can HyPrColoc jointly and quickly
analyse hundreds or thousands of traits? As we have seen from the
previous section, assessing the stability of clusters of putatively
colocalized traits is no simple task and this problem can become more
difficult as the number of traits under consideration increases. The
approach to assessing sensitivity of results to the choice of prior and
parameters, as well as evidential strength, is the same whether we
analyse 10 traits or 1000. Given this, here we focus on demonstrating
the ability of HyPrColoc to identify clusters of colocalized traits from
100 then 1000 traits. We do this by replicating the test dataset of 10
traits at first 10 times: to build a dataset of 100 traits in which
traits 1-50 form a cluster of colocalized traits, traits 51-80 form a
second cluster of colocalized traits and finally traits 81-100 form a
third cluster of colocalized traits, e.g.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 100 traits</span></span>
<span><span class="co"># beta100 replicates the betas for clusters 1, 2 and 3 10 times</span></span>
<span><span class="va">traits100</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"T"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">betas100</span> <span class="op">&lt;-</span> <span class="va">ses100</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">clst.traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">6</span><span class="op">:</span><span class="fl">8</span>, <span class="fl">9</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">times</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">c</span> <span class="kw">in</span> <span class="va">clusters</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">tmp.betas</span> <span class="op">&lt;-</span> <span class="va">tmp.ses</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="va">traits</span> <span class="op">&lt;-</span> <span class="va">clst.traits</span><span class="op">[[</span><span class="va">c</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">times</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tmp.betas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">tmp.betas</span>, <span class="va">betas</span><span class="op">[</span>, <span class="va">traits</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">tmp.ses</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">tmp.ses</span>, <span class="va">ses</span><span class="op">[</span>, <span class="va">traits</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">betas100</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">betas100</span>, <span class="va">tmp.betas</span><span class="op">)</span></span>
<span>  <span class="va">ses100</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ses100</span>, <span class="va">tmp.ses</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas100</span>, <span class="va">ses100</span>, trait.names <span class="op">=</span> <span class="va">traits100</span>, snp.id <span class="op">=</span> <span class="va">rsid</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ptm</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;    0.80    0.09    0.91</span></span>
<span><span class="co"># print the number of traits in each cluster</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">traits</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">c</span> <span class="kw">in</span> <span class="va">clusters</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">traits</span><span class="op">[</span><span class="va">c</span><span class="op">]</span>, split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">head.traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">traits</span><span class="op">)</span></span>
<span>  <span class="va">tail.traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">traits</span><span class="op">)</span></span>
<span>  <span class="co"># print the head and tail of the traits in each cluster</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cluster "</span>, <span class="va">c</span>, <span class="st">" contains "</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">traits</span><span class="op">)</span>, <span class="st">" traits"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"#####"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">head.traits</span>, <span class="st">"..."</span>, <span class="va">tail.traits</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"#####"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; [1] "cluster 1 contains 50 traits"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt;  [1] "T1"  "T2"  "T3"  "T4"  "T5"  "T6"  "..." "T45" "T46" "T47" "T48" "T49"</span></span>
<span><span class="co">#&gt; [13] "T50"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt; [1] "cluster 2 contains 30 traits"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt;  [1] "T51" "T52" "T53" "T54" "T55" "T56" "..." "T75" "T76" "T77" "T78" "T79"</span></span>
<span><span class="co">#&gt; [13] "T80"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt; [1] "cluster 3 contains 20 traits"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt;  [1] "T81"  "T82"  "T83"  "T84"  "T85"  "T86"  "..."  "T95"  "T96"  "T97" </span></span>
<span><span class="co">#&gt; [11] "T98"  "T99"  "T100"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span></code></pre></div>
<p>There are several points to note: (i) HyPrColoc correctly identified
the three clusters of colocalized traits; (ii) by borrowing strength
across more (colocalized) traits, the posterior probability explained by
each of the candidate snps has increased meaning we have incresed
confidence that we have identified the underlying causal variant (or one
in perfect LD with it) and; (iii) HyPrColoc did this in under 1 second.
Let’s do the same thing for 1000 traits,</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1000 traits</span></span>
<span><span class="va">traits1000</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"T"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">betas1000</span> <span class="op">&lt;-</span> <span class="va">ses1000</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">clst.traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">6</span><span class="op">:</span><span class="fl">8</span>, <span class="fl">9</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">times</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">c</span> <span class="kw">in</span> <span class="va">clusters</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">tmp.betas</span> <span class="op">&lt;-</span> <span class="va">tmp.ses</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="va">traits</span> <span class="op">&lt;-</span> <span class="va">clst.traits</span><span class="op">[[</span><span class="va">c</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">times</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tmp.betas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">tmp.betas</span>, <span class="va">betas</span><span class="op">[</span>, <span class="va">traits</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">tmp.ses</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">tmp.ses</span>, <span class="va">ses</span><span class="op">[</span>, <span class="va">traits</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">betas1000</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">betas1000</span>, <span class="va">tmp.betas</span><span class="op">)</span></span>
<span>  <span class="va">ses1000</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ses1000</span>, <span class="va">tmp.ses</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas1000</span>, <span class="va">ses1000</span>, trait.names <span class="op">=</span> <span class="va">traits1000</span>, snp.id <span class="op">=</span> <span class="va">rsid</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ptm</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   52.29    5.19   57.75</span></span>
<span><span class="co"># print the number of traits in each cluster</span></span>
<span><span class="co"># print the number of traits in each cluster</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">traits</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">c</span> <span class="kw">in</span> <span class="va">clusters</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">traits</span><span class="op">[</span><span class="va">c</span><span class="op">]</span>, split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">head.traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">traits</span><span class="op">)</span></span>
<span>  <span class="va">tail.traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">traits</span><span class="op">)</span></span>
<span>  <span class="co"># print the head and tail of the traits in each cluster</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cluster "</span>, <span class="va">c</span>, <span class="st">" contains "</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">traits</span><span class="op">)</span>, <span class="st">" traits"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"#####"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">head.traits</span>, <span class="st">"..."</span>, <span class="va">tail.traits</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"#####"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; [1] "cluster 1 contains 500 traits"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt;  [1] "T1"   "T2"   "T3"   "T4"   "T5"   "T6"   "..."  "T495" "T496" "T497"</span></span>
<span><span class="co">#&gt; [11] "T498" "T499" "T500"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt; [1] "cluster 2 contains 300 traits"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt;  [1] "T501" "T502" "T503" "T504" "T505" "T506" "..."  "T795" "T796" "T797"</span></span>
<span><span class="co">#&gt; [11] "T798" "T799" "T800"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt; [1] "cluster 3 contains 200 traits"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span>
<span><span class="co">#&gt;  [1] "T801"  "T802"  "T803"  "T804"  "T805"  "T806"  "..."   "T995"  "T996" </span></span>
<span><span class="co">#&gt; [10] "T997"  "T998"  "T999"  "T1000"</span></span>
<span><span class="co">#&gt; [1] "#####"</span></span></code></pre></div>
<p>Again, for 1000 traits HyPrColoc correctly identified the three
clusters of colocalized traits. This time, however, it took longer to do
so (still under 1 minute). The length of time the algorithm takes
depends on the complexity of the clustering between the traits. It is
therefore very much dependent on the region and traits under
consideration.</p>
<p>Notice that the posterior probability of colocalization has dropped
quite a lot for clusters 2 and 3 when analysing 1000 traits. This is a
feature of assessing large numbers of traits jointly in a Bayesian
framework, we discuss this nuance in more detail in the HyPrColoc text.
As a final point of note, we notice that the posterior probability
explained by the candidate snps in clusters 2 and 3 is ‘maximized’. That
is, we know from our previous assessment of the credible sets of snps,
in both clusters, that the candidate snp for cluster 2 is in perfect LD
with one other snp. In cluster 3, the candidate snp is in perfect LD
with four other snps. Hence, the maximum evidential support for a single
causal snp in clusters 2 and 3 can be 50% and 20% respectively, which is
what we identify in our analysis.</p>
</div>
<div class="section level2">
<h2 id="analysing-correlated-traits">Analysing correlated traits<a class="anchor" aria-label="anchor" href="#analysing-correlated-traits"></a>
</h2>
<p>When interest lies in identifying colocalization across a set of
traits whose summary data are correlated (due to overlapping
participants) three additional pieces of information are required: (i) a
matrix containing the putative correlation between the traits; (ii) a
matrix containing the LD between the snps in the region and; (iii) a
matrix containing the proportion of shared participants between each
pair of studies. Note the matrix of overlapping sample proportions has
diagonal elements set to 1 and off diagonal elements <span class="math inline">\((i,j)\)</span> denote the proportion of overlap
between the <span class="math inline">\(i\)</span>’th and <span class="math inline">\(j\)</span>’th studies, i.e. if N denotes the total
number of overlapping participants and <span class="math inline">\(\{N_{i},N_{j}\}\)</span> denote the numbers of
participants in the <span class="math inline">\(i\)</span>’th and <span class="math inline">\(j\)</span>’th studies respectively, then <span class="math inline">\(N/\left(N_{i}N_{j}\right)\)</span> is the <span class="math inline">\((i,j)=(j,i)\)</span> element. As sample overlap
information may not always be available, HyPrColoc defaults to assumes
all studies have the same participants.</p>
<div class="section level4">
<h4 id="important-before-going-further">Important, before going further…<a class="anchor" aria-label="anchor" href="#important-before-going-further"></a>
</h4>
<p>Question: should analyses be adjusted for observed correlation
between study summary data?</p>
<p>It is our recommendation that users first consider whether adjusting
for observed correlation between summary data is necessary. We here
detail several points to consider.</p>
<p>Traits which are strongly correlated are in general more likely to
colocalize. This has two important consequences: (i) whilst special
exceptions exist, generally the a-priori probability of colocalization
between two strongly correlated traits is necessarily larger than two
traits that are weakly correlated, analyses which ignore this are less
likely to capture the largest set of traits which are truley colocalized
(see the HyPrColoc text for more on this) and consequently; (ii) if a
subset of the cluster of traits which truley colocalize are identified,
it is possible (moreover likely) that one or more of the correlated
traits has been wrongly removed from the cluster due to mathematical
subtleties when inverting the correlation matrix (to estimate the
posterior probability of colocalization), and not because of some
underlying biological importance that one of the traits has over the
another.</p>
<p>As something of a discussion point and an example of (ii), let’s
suppose interest lies in assessing diastolic blood pressure (DBP) and
systolic blood pressure (SBP), two strongly correlated traits, and
coronary heart disease. Suppose we perform a joint analysis of the three
traits which accounts for correlation in the summary data, but not
a-priori trait correlation. Suppose further that we identify a region in
which SBP and CHD colocalize only. What do we conclude? Do we assume
that our findings provide evidence toward SBP being the more important
blood pressure trait for CHD risk in this region? Now suppose we
question this result by performing two separate colocalization analyses
(i) including SBP and CHD only and; (ii) including DBP and CHD only. Our
findings reveal that in this region CHD colocalizes with SBP and
separately colocalizes with DBP at the same snp. What then is the
conclusion? Is there a more important blood pressure trait here? Well,
we demonstrate in the HyPrColoc supplementary material that if analyses
account for ‘known’, i.e. a-priori, trait correlation by upweighting the
prior probability of colocalization between strongly correlated traits
as well as accounting for observed correlation, through the summary
effects, HyPrColoc regularly identifies the full set of truly
colocalized traits. So, in the case of SBP, DBP and CHD, additionally
accounting for ‘known’ correlation between SBP and DBP might overcome
this (hypothetical) issue. The difficulty with this approach is
identifying a robust way to incorporate ‘known’ trait correlation into
analyses. This is an outstanding problem. It appears it can be avoided
however, by ignoring both observed and a-priori correlations in
analyses, i.e. assuming independence between the studies and the traits.
We demonstrate that this strategy regularly identifies the maximum
number of truly colocalized traits and at a fraction of the
computational cost of analyses which account for correlation.</p>
</div>
<div class="section level4">
<h4 id="analysing-correlated-traits-continued">Analysing correlated traits, continued…<a class="anchor" aria-label="anchor" href="#analysing-correlated-traits-continued"></a>
</h4>
<p>This requires the user to enter trait correlation (using a
tetrachoric correlation technique) and LD matrices.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># default assumption, presented for clarity:</span></span>
<span><span class="va">sample.overlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">betas</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">betas</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="section level5">
<h5 id="variant-specific-priors">Variant specific priors<a class="anchor" aria-label="anchor" href="#variant-specific-priors"></a>
</h5>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"T"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">betas</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>,</span>
<span>  trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, trait.cor <span class="op">=</span> <span class="va">trait.cor</span>,</span>
<span>  ld.matrix <span class="op">=</span> <span class="va">ld.matrix</span>, sample.overlap <span class="op">=</span> <span class="va">sample.overlap</span>, uniform.priors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">time.corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ptm</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.8257             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.8944             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.3535            NA</span></span>
<span><span class="co">#&gt; 3                     0.0696            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="conditionally-uniform-priors">Conditionally uniform priors<a class="anchor" aria-label="anchor" href="#conditionally-uniform-priors"></a>
</h5>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>,</span>
<span>  trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span>, trait.cor <span class="op">=</span> <span class="va">trait.cor</span>,</span>
<span>  ld.matrix <span class="op">=</span> <span class="va">ld.matrix</span>, sample.overlap <span class="op">=</span> <span class="va">sample.overlap</span>, uniform.priors <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; $results</span></span>
<span><span class="co">#&gt;   iteration             traits posterior_prob regional_prob candidate_snp</span></span>
<span><span class="co">#&gt; 1         1 T1, T2, T3, T4, T5         1.0000             1    rs11591147</span></span>
<span><span class="co">#&gt; 2         2         T6, T7, T8         0.9744             1    rs12117612</span></span>
<span><span class="co">#&gt; 3         3            T9, T10         0.9782             1     rs7524677</span></span>
<span><span class="co">#&gt;   posterior_explained_by_snp dropped_trait</span></span>
<span><span class="co">#&gt; 1                     1.0000            NA</span></span>
<span><span class="co">#&gt; 2                     0.3535            NA</span></span>
<span><span class="co">#&gt; 3                     0.0696            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "hyprcoloc"</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="assuming-independence-which-correctly-captures-the-data-generating-model">Assuming independence (which correctly captures the data generating
model)<a class="anchor" aria-label="anchor" href="#assuming-independence-which-correctly-captures-the-data-generating-model"></a>
</h5>
<p>Recall that when we incorrectly assumed that the traits were measured
in studies with no overlapping participants, we correctly detected that
traits 1-5; traits 6-8 and traits 9-10 form three clusters of
colocalized traits and moreover we identified this at a fraction of the
computational cost of accounting for correlations between summary
data</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hyprcoloc.html">hyprcoloc</a></span><span class="op">(</span><span class="va">betas</span>, <span class="va">ses</span>, trait.names <span class="op">=</span> <span class="va">traits</span>, snp.id <span class="op">=</span> <span class="va">rsid</span><span class="op">)</span></span>
<span><span class="va">time.ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ptm</span></span>
<span><span class="va">time.ind</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;    0.03    0.00    0.03</span></span>
<span><span class="va">time.corr</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   33.97    0.23   34.28</span></span></code></pre></div>
<p>Thus, assuming independence between the studies not only correctly
identifies the three clusters of colocalized traits, but does so around
3500 times faster than accounting for non-indepdence.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Christopher Foley, James Staley.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
